### ISSUE TITLE
undo and validate_edit callback

### ISSUE THREAD

**Issue by skedus** [ragardner/tksheet#256]
Hi,
For each cell edition I need to process an external action, the validate_edit callback is working fine for it for most edition (perfect :-) ).
Now I try to manage the "undo" it seems that the validate_edit is not use. I tried the documented command:
```        
self.sheet.edit_validation(self.validate_edits).bind("<<SheetModified>>", self.sheet_modified)
``` 
during undo the self.sheet_modified is well call but I have no indication of the cells modified with the new values (only old values)

I tried as well extra_binding end_undo but this is the same event as self.sheet_modified

**Is there a way to manage each cell undo such validate_edit, with the cell row,col and new value of the cell ?**
Thank you

Here below the simple code I use to display event.
1 - I modify a cell
2 - Ctr-z to make the undo

```
from tksheet import Sheet
import tkinter as tk
import pprint


class demo(tk.Tk):
    def __init__(self):
        tk.Tk.__init__(self)
        self.grid_columnconfigure(0, weight=1)
        self.grid_rowconfigure(0, weight=1)
        self.frame = tk.Frame(self)
        self.frame.grid_columnconfigure(0, weight=1)
        self.frame.grid_rowconfigure(0, weight=1)

        self.sheet = Sheet(
            self.frame,
            data=[[f"Row {r}, Column {c}" for c in range(6)] for r in range(21)],
            column_width=180,
            height=700,
            width=1100,
        )
        self.sheet.enable_bindings(
            "edit_cell",
            "copy",
            "paste",
            "cut",
            "undo",
            "delete",
            "rc_insert_row",
            "rc_delete_row",
            "rc_select",
            "arrowkeys",
            "column_width_resize",
            "row_select",
            "single_select",
        )

        for row in range(self.sheet.get_total_rows()):
            self.sheet.create_checkbox(
                r=row,
                c=1,
                checked=False,
                check_function=lambda event=None: self.click_checkbox(event),
            )

        self.frame.grid(row=0, column=0, sticky="nswe")
        self.sheet.grid(row=0, column=0, sticky="nswe")

        self.sheet.edit_validation(self.validate_edits).bind("<<SheetModified>>", self.sheet_modified)
        self.sheet.extra_bindings("end_undo", self.end_undo)

    def click_checkbox(self, event):
        """Custom checkbox"""
        print(f"[click_checkbox] eventname={event.eventname} row={event.row} col={event.column} value={event.value}")

    def end_undo(self, event):
        print(f"[end_undo] event=")
        pprint.pp(event)

    def validate_edits(self, event):
        print(f"[validate_edits] event=")
        pprint.pp(event)
        return event.value

    def sheet_modified(self, event):
        print(f"[sheet_modified] event=")
        pprint.pp(event)


app = demo()
app.mainloop()

```

**Comment by ragardner** [comment]
Hello,

Thanks for your thoughts and your example,

I have decided to add edit validation checks to undo / redo for cell edits. The logic is basically that undo and redo are user actions, in the same way that cut, delete and paste are and perhaps there has been changes to the validation since the action or something

The work should be completed this week sometime

About the old values being in the event data - 

If its not an edit validation bound event, but a post action event such as with the `extra_bindings` you've got in your example you should be able to use the coordinates of the cell edits in the event data to get the new values from the sheet

Kind regards

**Comment by ragardner** [comment]
Hey,

I have made the following changes in `7.2.23`:

#### Changed:
- Edit validation will now also be triggered for undo and redo [256](https://github.com/ragardner/tksheet/issues/256)
- Cell text editors and dropdowns have their own shared background and foreground colors [255](https://github.com/ragardner/tksheet/issues/255)

#### Fixed:
- `7.2.22` regression when using cell text editors and normal dropdown boxes
- `7.2.22` regression `"normal"` dropdown box with `modified_function` not being triggered upon non-key release events such as by right clicking in the text editor and pasting
- Select events will no longer be emitted upon b1 release

#### Added:
- Options for changing table, index and header cell editor bg, fg [255](https://github.com/ragardner/tksheet/issues/255)
    - Added `Sheet()` parameters:
        - `table_editor_bg`
        - `table_editor_fg`
        - `table_editor_select_bg`
        - `table_editor_select_fg`
        - `index_editor_bg`
        - `index_editor_fg`
        - `index_editor_select_bg`
        - `index_editor_select_fg`
        - `header_editor_bg`
        - `header_editor_fg`
        - `header_editor_select_bg`
        - `header_editor_select_fg`

Let me know if you have any issues with the above!

**Comment by skedus** [comment]
Thank you very much ragardner, I will test it very soon and give you some feedback on it

**Comment by skedus** [comment]
The undo is working for a cell event.row event.column and event.value is correctly set into validate_edit
So this update solve the problem. Thanks :-)

More in details event in the validate_edit is a little bit different for an edit and an undo:

Example Modification of cell(3,2) to "aaaa" then undo

**Edition:** 
'eventname': 'end_edit_table'
'cells': {'table': {(3, 2): 'aaaa'}, 'header': {}, 'index': {}},
'row': 3,
'column': 2,
'value': 'aaaa',

**Undo:**
'eventname': 'edit_table'
'cells': {'table': {(3, 2): 'aaaa'}, 'header': {}, 'index': {}}
'row': 3,
'column': 2,
'value': 'Row 3, Column 2',

During undo the cells: contain old value the eventname is edit_table
I imagine one event is done at the end of edition the other at the begining, I do not know if it is good or not and it is not a problem in my case as row,column, and value are correct. it is just a remark (because validation may be done more at the end like edition ?)

Thank you for your support

**Comment by ragardner** [comment]
Thank you for letting me know about the difference in values under `cells`.

I think that I will have to change the values under `cells` for the `end...` events to be the cell values prior to the change for consistency and to adhere to the existing documentation.

Considering it's been like this a while and there is the possibility that some people may be using these values I think I will have to release it with version `7.3.x`.

**Comment by ragardner** [comment]
I have released version `7.3.0` which changes the `"end_edit_table"` event `cells.table` value to the value prior to the change.

This is now in line with other such events as selecting an item in a dropdown or clicking a checkbox, also in the index and header.

Once again thanks for your reports and let me know if you have further issues!

Kind regards

**Comment by skedus** [comment]
Perfect Thank you very much
